File: examples/pmai-task-prioritizer/pipeline/runner.py
"""Pipeline runner: executes the six stages."""
import asyncio
import logging
from typing import Generic, TypeVar

from .interfaces import (
    Source, Hydrator, Filter, Scorer, Selector, SideEffect, ScoredItem,
)

Ctx = TypeVar("Ctx")
Item = TypeVar("Item")

log = logging.getLogger(__name__)


class Pipeline(Generic[Ctx, Item]):
    def __init__(
        self,
        sources: list[Source[Ctx, Item]],
        hydrators: list[Hydrator[Ctx, Item]],
        filters: list[Filter[Ctx, Item]],
        scorers: list[Scorer[Ctx, Item]],
        selector: Selector[Ctx, Item],
        side_effects: list[SideEffect[Ctx, Item]],
    ):
        self.sources = sources
        self.hydrators = hydrators
        self.filters = filters
        self.scorers = scorers
        self.selector = selector
        self.side_effects = side_effects

    async def run(self, ctx: Ctx, k: int) -> list[ScoredItem[Item]]:
        # 1. Sources parallel
        source_results = await asyncio.gather(
            *[self._safe_fetch(s, ctx) for s in self.sources]
        )
        items: list[Item] = [it for batch in source_results for it in batch]

        # 2. Hydrators parallel
        await asyncio.gather(
            *[self._safe_hydrate(h, ctx, items) for h in self.hydrators]
        )

        # 3. Filters sequential
        for f in self.filters:
            items = [it for it in items if f.keep(ctx, it)]

        # 4. Scorers sequential
        scored: list[ScoredItem[Item]] = [ScoredItem(item=it) for it in items]
        for s in self.scorers:
            await s.score(ctx, scored)

        # 5. Selector
        selected = self.selector.select(ctx, scored, k)

        # 6. Side effects fire-and-forget
        for se in self.side_effects:
            asyncio.create_task(self._safe_emit(se, ctx, selected))

        return selected

    async def _safe_fetch(self, s: Source[Ctx, Item], ctx: Ctx) -> list[Item]:
        try:
            return await s.fetch(ctx)
        except Exception as e:
            log.warning("Source %s failed: %s", s.name, e)
            return []

    async def _safe_hydrate(self, h: Hydrator[Ctx, Item], ctx: Ctx, items: list[Item]) -> None:
        try:
            await h.hydrate(ctx, items)
        except Exception as e:
            log.warning("Hydrator %s failed: %s", h.name, e)

    async def _safe_emit(self, se: SideEffect[Ctx, Item], ctx: Ctx, items: list[ScoredItem[Item]]) -> None:
        try:
            await se.emit(ctx, items)
        except Exception as e:
            log.warning("SideEffect %s failed: %s", se.name, e)

File: examples/strapi-content-feed/src/pipeline/runner.ts
import {
  Source, Hydrator, Filter, Scorer, Selector, SideEffect, ScoredItem,
} from './interfaces';

export interface PipelineConfig<Ctx, Item> {
  sources: Source<Ctx, Item>[];
  hydrators: Hydrator<Ctx, Item>[];
  filters: Filter<Ctx, Item>[];
  scorers: Scorer<Ctx, Item>[];
  selector: Selector<Ctx, Item>;
  sideEffects: SideEffect<Ctx, Item>[];
}

export class Pipeline<Ctx, Item> {
  constructor(private config: PipelineConfig<Ctx, Item>) {}

  async run(ctx: Ctx, k: number): Promise<ScoredItem<Item>[]> {
    // 1. Sources in parallel
    const sourceResults = await Promise.all(
      this.config.sources.map(s => s.fetch(ctx).catch(err => {
        console.error(`Source ${s.name} failed:`, err);
        return [] as Item[];
      }))
    );
    let items: Item[] = sourceResults.flat();

    // 2. Hydrators in parallel (each enriches the full set)
    await Promise.all(
      this.config.hydrators.map(h =>
        h.hydrate(ctx, items).catch(err => {
          console.error(`Hydrator ${h.name} failed:`, err);
        })
      )
    );

    // 3. Filters sequentially
    for (const f of this.config.filters) {
      const checks = await Promise.all(
        items.map(item => Promise.resolve(f.keep(ctx, item)))
      );
      items = items.filter((_, i) => checks[i]);
    }

    // 4. Scorers sequentially (each sees previous scores)
    let scored: ScoredItem<Item>[] = items.map(item => ({
      item,
      score: 0,
      components: {},
    }));
    for (const s of this.config.scorers) {
      await s.score(ctx, scored);
    }

    // 5. Selector
    const selected = this.config.selector.select(ctx, scored, k);

    // 6. SideEffects fire-and-forget
    for (const se of this.config.sideEffects) {
      Promise.resolve(se.emit(ctx, selected)).catch(err => {
        console.error(`SideEffect ${se.name} failed:`, err);
      });
    }

    return selected;
  }
}
